keps/sig-auth/20190607-certificates-api.md

```diff
@@ -145,6 +227,22 @@ type CertificateSigningRequest struct {
 
 
 type CertificateSigningRequestSpec struct {
+  // requested signer for the request.  It is a qualified name in the form: `scope-hostname.io/name`.  
```

Comment:
@munnerz and I were discussing people might want to encode in a signerName... letting the `<name>` portion be long enough to contain `<namespace name>.<custom resource name>` would be helpful for referencing dynamic namespaced signers

Reply:
If we want to support namespaced signers, we will need to consider adding an authorization check for "sign" as well

Reply:
> If we want to support namespaced signers, we will need to consider adding an authorization check for "sign" as well

Do you mean an authz plugin that checks that a user has permission to *sign* a CSR, i.e. requiring extra RBAC permissions in order to set `status.certificate`?

This may well be useful, although I don't think it directly relates to 'namespaced signers' - the format @liggitt mentions above is, if anything, an implementation detail of cert-manager: how we encode a reference to what *we* call an 'Issuer' (i.e. a 'dynamically registered signer').

Whilst the representation of that signer can be namespaced in our project, because CSRs are *non-namespaced* the thinking was to use the signerName field to encode the kind, group, name and namespace of the 'issuer' that should be used to sign certificates.

Ultimately, the same cert-manager-controller still operates on these CSRs, i.e. the 'namespace' of the issuer does not imply that the controller signing those certs runs within that namespace - it represents a 'configuration' of one of cert-manager's built in signer types, serviced by the main cert-manager-controller (or potentially a different controller, but the key point is that it is a singleton in the cluster for each GroupKind).

From an end-user perspective, the most awkward part right now is that in a multi-tenant cluster where they have configured their own namespaced issuer resource, they will *not* be able to interact with CertificateSigningRequest resources directly as they could see other tenants own CSRs (i.e. running `kubectl get csr`). I don't think this *directly* relates though, and the `sign` verb wouldn' t really help here either as the issue is really information leakage (i.e. one tenant can work out what the other tenant is doing).

Reply:
> I don't think it directly relates to 'namespaced signers' - the format @liggitt mentions above is, if anything, an implementation detail of cert-manager: how we encode a reference to what _we_ call an 'Issuer' (i.e. a 'dynamically registered signer').

Yeah, if someone actually had a namespaced signer implementation, they would be responsible for translating/delegating CSR objects to/from their namespaced signers. I just wanted to avoid making someone have to do truncated+hashed suffixes of signer names or something really ugly to represent `namespaced-signers.acme.com/<namespace>.<signer>`

Reply:
> Do you mean an authz plugin that checks that a user has permission to sign a CSR, i.e. requiring extra RBAC permissions in order to set status.certificate?

I was considering this as well (separate from this signerName validation topic). If we're going to introduce signer-specific authorization, now's the time. If critical things like node client certs are running on CSRs, a signer accident could take down your cluster. It would mostly be for accident prevention from my perspective. I would imagine it being shaped the same way as `approve` subdivision, with essentially the same compatibility implications and options for cluster admins (grant bootstrap roles, grant custom roles, or disable the admission plugin). I'm slightly in favor of treating signing and approving consistently for purposes of authorization subdivision by signerName.

Reply:
>  I just wanted to avoid making someone have to do truncated+hashed suffixes of signer names

:+1: agreed - what is the minimum required set of things here to make this possible? Right now in the implementation PR there is no validation on the `signerName` field at all.

> If we're going to introduce signer-specific authorization, now's the time.

Agreed - although I'm uncertain whether treating them as the same thing today is a good move given it'll be difficult to untangle in future. Requiring one extra `verb` on the RBAC rule users create is fairly non-intrusive, and it allows for better segmentation of concerns for those with tighter security requirements.

That said, I do have some concerns around user configuration requirements with the current proposed RBAC model: https://github.com/kubernetes/enhancements/pull/1400#discussion_r363989483

There is no way to grant permission for an entity to approve (and potentially 'sign') *all* signers within their namespace, e.g. `kubernetes.io/*` because the `name` field of an RBAC rule cannot contain wildcards. This may make migrations for e.g. `kubernetes.io/*` signerNames that may be added in future more difficult, as well as requiring users registering dynamic signers (i.e. cert-manager users) to grant the cert-manager pod `sign` and `approve` permissions for any new Issuer or ClusterIssuer type they register. Perhaps this is a good thing, but it also represents a new amount of operational overhead, and given that a namespaced user cannot create `ClusterRole(Binding)` resources, requires cooperation with an administrator whenever a user wants to use cert-manager create a private PKI within a namespace.

Reply:
After discussion on Slack, @liggitt suggested that we could make two checks against the authorization API in order to determine whether a user has permission on either "foo.bar/baz", and failing that, check if they have permission for "foo.bar/*".

Whilst this isn't perfect, as the authorization system itself *won't* be able to understand the `*` (wildcard) in the resource name, it _is_ sufficient and should help at least unblock the above.

https://kubernetes.slack.com/archives/C0EN96KUY/p1578672804008900

Reply:
Regarding RBAC for *signing* permissions (similar to approval), it sounds like there's general agreement that this would be beneficial (https://kubernetes.slack.com/archives/C0EN96KUY/p1578675929056700)

Would this be implemented in a similar fashion to the `approve` verb, but as a `sign` or `issue` verb? Is there any real value in overloading the `approve` verb and using it for both?

Reply:
> Would this be implemented in a similar fashion to the approve verb, but as a sign or issue verb?

That would be my inclination

> Is there any real value in overloading the approve verb and using it for both?

We would not do this. Approving for a signer and signing for that signer are two different things.